Bottom: 51864b089669e4e87a1c4ca0cf82050f6bad16d0
Top:    4e7eadfc2eca9275d876db1d11f37926a8292ed9
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-10-21 00:00:12 +0100

Refresh of major-re-write-to-drawing-rout

---

diff --git a/src/hid/gtk/gui-output-events.c b/src/hid/gtk/gui-output-events.c
index 325b7b4..39a65c9 100644
--- a/src/hid/gtk/gui-output-events.c
+++ b/src/hid/gtk/gui-output-events.c
@@ -1484,6 +1484,7 @@ ghid_draw_everything (BoxTypePtr drawn_area)
   /* This is the reverse of the order in which we draw them.  */
   int drawn_groups[MAX_LAYER];
   struct cyl_info cyl_info;
+  int reverse_layers;
 
   extern char *current_color;
   extern Boolean Gathering;
@@ -1491,12 +1492,29 @@ ghid_draw_everything (BoxTypePtr drawn_area)
   current_color = NULL;
   Gathering = False;
 
+  /* Test direction of rendering */
+  /* Look at sign of eye coordinate system z-coord when projecting a
+     world vector along +ve Z axis, (0, 0, 1). */
+  /* FIXME: This isn't strictly correct, as I've ignored the matrix
+            elements for homogeneous coordinates. */
+  /* NB: last_modelview_matrix is transposed in memory! */
+  reverse_layers = (last_modelview_matrix[2][2] < 0);
+  if (Settings.ShowSolderSide)
+    reverse_layers = !reverse_layers;
+
+  Settings.ShowSolderSide = reverse_layers ? !Settings.ShowSolderSide : Settings.ShowSolderSide;
+
   memset (do_group, 0, sizeof (do_group));
   for (ngroups = 0, i = 0; i < max_layer; i++) {
+    LayerType *l;
+    int group;
+    int orderi;
+
+    orderi = reverse_layers ? max_layer - i - 1 : i;
 
-    // Draw in numerical order
-    LayerType *l = LAYER_PTR (i); // LAYER_ON_STACK (i);
-    int group = GetLayerGroupNumberByNumber (i); // (LayerStack[i]);
+    // Draw in numerical order when in 3D view
+    l = global_view_2d ? LAYER_ON_STACK (i) : LAYER_PTR (orderi);
+    group = GetLayerGroupNumberByNumber (global_view_2d ? LayerStack[i] : orderi);
 
     if (/*l->On && */!do_group[group]) {
       do_group[group] = 1;
@@ -1599,6 +1617,7 @@ ghid_draw_everything (BoxTypePtr drawn_area)
     DrawRats(drawn_area);
 
   Gathering = True;
+  Settings.ShowSolderSide = reverse_layers ? !Settings.ShowSolderSide : Settings.ShowSolderSide;
 }
 
 static int one_shot = 1;
